File: windows/LiteDesk/Native/KeyCodeMap.cs
namespace LiteDesk.Native;

// Translates between Win32 virtual-key codes (winuser.h VK_* constants) and
// the layout-independent DOM `KeyboardEvent.code` strings used on the wire
// (see Protocol/WireMessages.cs). The same code string space is shared with
// the macOS app's KeyCodeMap.swift, so either side can send a code the other
// side has never seen a native keycode for.
internal static class KeyCodeMap
{
    internal static string? DomCode(int virtualKey)
    {
        return VkToDomCode.TryGetValue(virtualKey, out string? code) ? code : null;
    }

    internal static ushort? VirtualKey(string code)
    {
        return DomCodeToVk.TryGetValue(code, out ushort vk) ? vk : null;
    }

    private static readonly Dictionary<int, string> VkToDomCode = new()
    {
        [0x41] = "KeyA", [0x42] = "KeyB", [0x43] = "KeyC", [0x44] = "KeyD", [0x45] = "KeyE",
        [0x46] = "KeyF", [0x47] = "KeyG", [0x48] = "KeyH", [0x49] = "KeyI", [0x4A] = "KeyJ",
        [0x4B] = "KeyK", [0x4C] = "KeyL", [0x4D] = "KeyM", [0x4E] = "KeyN", [0x4F] = "KeyO",
        [0x50] = "KeyP", [0x51] = "KeyQ", [0x52] = "KeyR", [0x53] = "KeyS", [0x54] = "KeyT",
        [0x55] = "KeyU", [0x56] = "KeyV", [0x57] = "KeyW", [0x58] = "KeyX", [0x59] = "KeyY",
        [0x5A] = "KeyZ",

        [0x30] = "Digit0", [0x31] = "Digit1", [0x32] = "Digit2", [0x33] = "Digit3", [0x34] = "Digit4",
        [0x35] = "Digit5", [0x36] = "Digit6", [0x37] = "Digit7", [0x38] = "Digit8", [0x39] = "Digit9",

        [0x70] = "F1", [0x71] = "F2", [0x72] = "F3", [0x73] = "F4", [0x74] = "F5", [0x75] = "F6",
        [0x76] = "F7", [0x77] = "F8", [0x78] = "F9", [0x79] = "F10", [0x7A] = "F11", [0x7B] = "F12",

        [0x0D] = "Enter", [0x09] = "Tab", [0x20] = "Space", [0x1B] = "Escape", [0x08] = "Backspace",
        [0x2E] = "Delete", [0x2D] = "Insert", [0x24] = "Home", [0x23] = "End",
        [0x21] = "PageUp", [0x22] = "PageDown",
        [0x25] = "ArrowLeft", [0x26] = "ArrowUp", [0x27] = "ArrowRight", [0x28] = "ArrowDown",
        [0x14] = "CapsLock", [0x90] = "NumLock", [0x91] = "ScrollLock",

        [0xA0] = "ShiftLeft", [0xA1] = "ShiftRight",
        [0xA2] = "ControlLeft", [0xA3] = "ControlRight",
        [0xA4] = "AltLeft", [0xA5] = "AltRight",
        [0x5B] = "MetaLeft", [0x5C] = "MetaRight",

        [0xBD] = "Minus", [0xBB] = "Equal", [0xDB] = "BracketLeft", [0xDD] = "BracketRight",
        [0xDC] = "Backslash", [0xBA] = "Semicolon", [0xDE] = "Quote",
        [0xBC] = "Comma", [0xBE] = "Period", [0xBF] = "Slash", [0xC0] = "Backquote",

        [0x60] = "Numpad0", [0x61] = "Numpad1", [0x62] = "Numpad2", [0x63] = "Numpad3",
        [0x64] = "Numpad4", [0x65] = "Numpad5", [0x66] = "Numpad6", [0x67] = "Numpad7",
        [0x68] = "Numpad8", [0x69] = "Numpad9",
        [0x6A] = "NumpadMultiply", [0x6B] = "NumpadAdd", [0x6D] = "NumpadSubtract",
        [0x6E] = "NumpadDecimal", [0x6F] = "NumpadDivide",
    };

    private static readonly Dictionary<string, ushort> DomCodeToVk =
        VkToDomCode.ToDictionary(pair => pair.Value, pair => (ushort)pair.Key);
}

File: macos/LiteDesk/Input/KeyCodeMap.swift
import CoreGraphics

/// Translates between macOS's native virtual keycodes (CGKeyCode / NSEvent.keyCode,
/// per Carbon's HIToolbox Events.h kVK_* constants) and the layout-independent DOM
/// `KeyboardEvent.code` strings used on the wire (see WireMessages.swift). The same
/// code string space is shared with the Windows app's KeyCodeMap.cs, so either side
/// can send a code the other side has never seen a native keycode for.
enum KeyCodeMap {
    static func domCode(forKeyCode keyCode: UInt16) -> String? {
        keyCodeToDomCode[keyCode]
    }

    static func keyCode(forDomCode code: String) -> CGKeyCode? {
        domCodeToKeyCode[code]
    }

    private static let keyCodeToDomCode: [UInt16: String] = [
        0x00: "KeyA", 0x01: "KeyS", 0x02: "KeyD", 0x03: "KeyF", 0x04: "KeyH",
        0x05: "KeyG", 0x06: "KeyZ", 0x07: "KeyX", 0x08: "KeyC", 0x09: "KeyV",
        0x0B: "KeyB", 0x0C: "KeyQ", 0x0D: "KeyW", 0x0E: "KeyE", 0x0F: "KeyR",
        0x10: "KeyY", 0x11: "KeyT", 0x12: "Digit1", 0x13: "Digit2", 0x14: "Digit3",
        0x15: "Digit4", 0x16: "Digit6", 0x17: "Digit5", 0x18: "Equal", 0x19: "Digit9",
        0x1A: "Digit7", 0x1B: "Minus", 0x1C: "Digit8", 0x1D: "Digit0", 0x1E: "BracketRight",
        0x1F: "KeyO", 0x20: "KeyU", 0x21: "BracketLeft", 0x22: "KeyI", 0x23: "KeyP",
        0x24: "Enter", 0x25: "KeyL", 0x26: "KeyJ", 0x27: "Quote", 0x28: "KeyK",
        0x29: "Semicolon", 0x2A: "Backslash", 0x2B: "Comma", 0x2C: "Slash", 0x2D: "KeyN",
        0x2E: "KeyM", 0x2F: "Period", 0x30: "Tab", 0x31: "Space", 0x32: "Backquote",
        0x33: "Backspace", 0x35: "Escape",
        0x37: "MetaLeft", 0x38: "ShiftLeft", 0x39: "CapsLock", 0x3A: "AltLeft",
        0x3B: "ControlLeft", 0x3C: "ShiftRight", 0x3D: "AltRight", 0x3E: "ControlRight",
        0x41: "NumpadDecimal", 0x43: "NumpadMultiply", 0x45: "NumpadAdd",
        0x4B: "NumpadDivide", 0x4C: "NumpadEnter", 0x4E: "NumpadSubtract",
        0x51: "NumpadEqual", 0x52: "Numpad0", 0x53: "Numpad1", 0x54: "Numpad2",
        0x55: "Numpad3", 0x56: "Numpad4", 0x57: "Numpad5", 0x58: "Numpad6",
        0x59: "Numpad7", 0x5B: "Numpad8", 0x5C: "Numpad9",
        0x60: "F5", 0x61: "F6", 0x62: "F7", 0x63: "F3", 0x64: "F8", 0x65: "F9",
        0x67: "F11", 0x69: "F13", 0x6A: "F16", 0x6B: "F14", 0x6D: "F10", 0x6F: "F12",
        0x71: "F15", 0x72: "Insert", 0x73: "Home", 0x74: "PageUp", 0x75: "Delete",
        0x76: "F4", 0x77: "End", 0x78: "F2", 0x79: "PageDown", 0x7A: "F1",
        0x7B: "ArrowLeft", 0x7C: "ArrowRight", 0x7D: "ArrowDown", 0x7E: "ArrowUp",
    ]

    private static let domCodeToKeyCode: [String: CGKeyCode] = {
        var reversed: [String: CGKeyCode] = [:]
        for (keyCode, code) in keyCodeToDomCode {
            reversed[code] = keyCode
        }
        return reversed
    }()
}
